File: packages/postgres/src/ddl.ts
import type { Queryable } from "./index.ts";

/**
 * Runs a DDL statement the way several servers starting at once need it run. `CREATE ... IF NOT EXISTS` is not atomic
 * between sessions: when two run it at the same moment, Postgres blocks the second until the first commits and then
 * refuses it with a duplicate key on its catalogue (`23505`) or "already exists" (`42P07`). The object exists by then,
 * which is all that was wanted, so the statement is run once more and passes.
 */
export async function ensure(sql: Queryable, statement: string): Promise<void> {
  try {
    await sql.query(statement);
  } catch (e) {
    const code = (e as { code?: unknown }).code;
    if (code !== "23505" && code !== "42P07") throw e;
    await sql.query(statement);
  }
}

File: kotlin/rayfold-jdbc/src/main/kotlin/dev/rayfold/jdbc/Ddl.kt
package dev.rayfold.jdbc

import java.sql.Connection
import java.sql.SQLException

/**
 * Runs a `CREATE ... IF NOT EXISTS` so that every server may run it as it starts. The statement is not atomic across
 * sessions: Postgres blocks the second until the first commits, then refuses it with 23505 (a duplicate key in
 * pg_type) or 42P07 (the relation already exists). The object exists by then, which is all that was wanted, so the
 * statement runs once more, which finds it there. Anything else propagates untouched.
 */
internal fun ensure(connection: Connection, ddl: String) {
    try {
        connection.createStatement().use { it.execute(ddl) }
    } catch (e: SQLException) {
        if (e.sqlState != "23505" && e.sqlState != "42P07") throw e
        connection.createStatement().use { it.execute(ddl) }
    }
}
